### ISSUE TITLE
Matching IP Family for Outbound Listener

### ISSUE THREAD

**Issue by ycai-aspen** [envoyproxy/envoy#16804]
*Title*: Matching IP Family for Outbound Listener

*Description*:
To support dual stack in istio, for each service that has dual stack configured, two outbound listeners for the same port are created. One listener is with ipv6 wildcard with port and the other one is with ipv4 wildcard with port. 

Here is what we expect:
client (ipv4 address ) -> ipv4 iptable redirect to 15001 -> envoy original dst at 15001 -> only match ipv4 wildcard 0.0.0.0:<port> listener
client (ipv6 address ) -> ipv6 iptable redirect to 15001 -> envoy original dst at 15001 -> only match ipv6 wildcard [::]:<port> listener

However, in the [connection_handler_impl.cc](https://github.com/envoyproxy/envoy/blob/main/source/server/connection_handler_impl.cc#L194-L205)
it was not matching IP family type for the listener and target address.  Instead, it finds the first wildcard matching port from the listener list and it always returns that listener. This causes downstream client can use a different IP family listener from the original destination address. We think the code should match IP family and pick the corresponding wildcard listener based on the original destination IP family type.

```
ConnectionHandlerImpl::findActiveTcpListenerByAddress(const Network::Address::In
                    p.second.typed_listener_) &&
                p.second.listener_->listener() != nullptr &&
                p.first->type() == Network::Address::Type::Ip &&
-               p.first->ip()->port() == address.ip()->port() && p.first->ip()->isAnyAddress();
+               p.first->ip()->port() == address.ip()->port() && p.first->ip()->isAnyAddress() &&
+              ((p.first->ip()->ipv4() != nullptr && address.ip()->ipv4() != nullptr) ||
+               (p.first->ip()->ipv6() != nullptr && address.ip()->ipv6() != nullptr));
       });
```

With the above change, we tested with success of either IPv4 or IPv6 path.
```
kubectl exec -it -n test sleep-86786f67f4-s5bfm sh
kubectl exec [POD] [COMMAND] is DEPRECATED and will be removed in a future version. Use kubectl exec [POD] -- [COMMAND] instead.
Defaulted container "sleep" out of: sleep, istio-proxy, istio-init (init)
/ # curl 10.96.65.101:8090/headers
X-Request-Id: a6b06ffb-724b-4785-b490-039dbde0baee
X-B3-Spanid: c1d76544d090db82
Accept: */*
X-Envoy-Attempt-Count: 1
Content-Length: 0
X-Forwarded-Client-Cert: By=spiffe://cluster.local/ns/test/sa/simpleserver;Hash=01b9728bd23645e1afc39c96826cbec8b46c6427ac9b147b998935cdd2271adc;Subject="";URI=spiffe://cluster.local/ns/test/sa/sleep
X-B3-Traceid: 24c6d5522b6172d88f843df561d424d8
X-B3-Parentspanid: 8f843df561d424d8
X-B3-Sampled: 0
User-Agent: curl/7.76.1
X-Forwarded-Proto: http
remote addr: 127.0.0.1:48372
host addr: 10.96.65.101:8090
/ # curl simpleserver:8090/headers
X-Forwarded-Client-Cert: By=spiffe://cluster.local/ns/test/sa/simpleserver;Hash=01b9728bd23645e1afc39c96826cbec8b46c6427ac9b147b998935cdd2271adc;Subject="";URI=spiffe://cluster.local/ns/test/sa/sleep
X-B3-Spanid: ea96060aa7025f33
X-B3-Parentspanid: f5e7ec00645e238a
User-Agent: curl/7.76.1
Accept: */*
X-Forwarded-Proto: http
X-Request-Id: f6e2dcb1-664d-4f54-baa1-0d5ca615a562
X-Envoy-Attempt-Count: 1
Content-Length: 0
X-B3-Traceid: 872552dfa9f3c99bf5e7ec00645e238a
X-B3-Sampled: 0
remote addr: [::1]:58248
host addr: simpleserver:8090
ipv6 only server
```
This is important for us to enable dual mode support in Istio as we create routes/clusters based on whether IPv4 or IPv6 listener is picked. 

*Repro steps*:
> Include sample requests, environment, etc. All data and inputs
required to reproduce the bug.

>**Note**: The [Envoy_collect tool](https://github.com/envoyproxy/envoy/blob/main/tools/envoy_collect/README.md)
gathers a tarball with debug logs, config and the following admin
endpoints: /stats, /clusters and /server_info. Please note if there are
privacy concerns, sanitize the data prior to sharing the tarball/pasting.
Attached the config_dump from istio.

*Admin and Stats Output*:
>Include the admin output for the following endpoints: /stats,
/clusters, /routes, /server_info. For more information, refer to the
[admin endpoint documentation.](https://www.envoyproxy.io/docs/envoy/latest/operations/admin)

*Config*:
>Include the config used to configure Envoy.

*Logs*:
>Envoy logs with added trace:

**Comment by ycai-aspen** [comment]
@nrjpoddar, @jacob-delgado

**Comment by nrjpoddar** [comment]
cc @howardjohn @mattklein123 @alyssawilk for inputs.

**Comment by lambdai** [comment]
Good point. I can make the change. 

BTW: Do you have a benchmark on the cost of the linear search? I have the feeling it hurts performance

**Comment by ycai-aspen** [comment]
We don't have a benchmark test on this.  It might be better to find the matching port listeners, then do a match on the ipaddress or wildcard with ip family.

**Comment by nrjpoddar** [comment]
Thanks @lambdai for your inputs. I was wondering about the linear search myself and I'm positive we can do better. However, I would like to get this bug fixed and tackle performance improvement separately if that's ok with you. Can we patch this in older versions of Envoy?

**Comment by jacob-delgado** [comment]
By older versions, @nrjpoddar is referring to Envoy 1.17 and Envoy 1.18. We wouldn't have to go further back than that. Once the PR is done I can help in cherry picking this into older releases.

I'm also of the opinion that we should fix this, then we can measure performance and see if we should optimize it as a separate issue. The amount of listeners in large systems could easily lead to performance bottlenecks if there exists one for both IPv4 and IPv6.

**Comment by nrjpoddar** [comment]
Great! @lambdai will you able to submit the fix in short order? We need this rather quickly as it blocks the dual-stack support in Istio. We would be happy to submit a patch too if you're busy.

**Comment by mattklein123** [comment]
This sounds reasonable to me. I might recommend putting this change behind a runtime flag though.

**Comment by nrjpoddar** [comment]
> This sounds reasonable to me. I might recommend putting this change behind a runtime flag though.

Thanks Matt! Why do you suggest adding a runtime flag? I can’t think of situations where the current implementation of first wildcard match would be preferred instead of deterministic IP family wildcard match.

**Comment by mattklein123** [comment]
> Thanks Matt! Why do you suggest adding a runtime flag? I can’t think of situations where the current implementation of first wildcard match would be preferred instead of deterministic IP family wildcard match.

I have no idea, but it strikes me as a change that could break someone. If you are sure this is impossible, we can proceed without it, but that seems risky to me as this is a behavior change to the matching.

**Comment by nrjpoddar** [comment]
Well, this falls into the category of users relying on incorrect existing behavior which is difficult to find out. At least in Istio, I know this won’t break anyone as we are still building out dual stack support. But since usage of Envoy is much broader, we can add a flag but make the new behavior default? I hope that’s acceptable.

**Comment by mattklein123** [comment]
Enabled by default is the default for runtime guards.
